A class and task that clones a case (#3826)

rsutphin opened the pull request:
Provides `setup:clone_case`, a rake task which produces one or more clones of a participant and all its related persons, participants, and other case-specific records. It also syncs any primary participants to PSC. This task is intended for use in setting up training or QA data.

This branch also includes a fix for [Cases-3848](https://code.bioinformatics.northwestern.edu/issues/issues/show/3848) because it involves a refactoring of `Participant` to share its "importer mode" code with `ResponseSet`.

rsutphin committed:
Handle cloning all Cases-side records for a case. #3826.

rsutphin committed:
Glue in PSC sync. #3826.

No tests because they'd just be a big mock mess. There's no logic here,
just wiring.

rsutphin committed:
Task for using CaseCloner from the CLI. #3826.

rsutphin committed:
Break loop CL -> C -> CL. #3826.

There are probably others like this, but not looking now.

rsutphin committed:
Refinements for PSC sync part. #3826.

* Correct keygen so that it is random for a run, not random for each key
* Don't save on the cases side if the PSC sync fails
* Debug logging

rsutphin committed:
Factor out "importer mode" flag & related. Done with #3826.

Also addresses #3848 along the way.

rsutphin committed:
Disable operational data extraction in import mode. #3826.

rsutphin committed:
Use importer_mode if available. #3826.

A clone is like an import in that you are looking to get an exact copy
of the source data. The same sort of record-creating callbacks that
are disabled for imports should be disabled for clones also.

rsutphin committed:
Add more case-specific models to clone. #3826.

hannahwhy commented on lib/ncs_navigator/core/import_aware.rb:
```diff
@@ -0,0 +1,29 @@
+        fail "This method is intended for use with a block" unless block_given?
+        original = importer_mode_on
+        self.importer_mode_on = true
+        result = yield
```
If the block raises an exception, `#importer_mode_on` won't be reset to its original value.  Is that intentional?

rsutphin replied:
No, that's not intentional, though for the applications this has had so far it isn't relevant. Worth fixing.

hannahwhy commented on lib/tasks/setup.rake:
```diff
@@ -32,4 +32,26 @@ namespace :setup do
+    psc_user = task('import:psc_setup').user
+    p_id = args[:p_id] or fail "Please specify a p_id: rake #{t.name}[12-1234]"
+    n = args[:n].try(:to_i) || 1
+    digits = (Math.log(n, 10) + 1).to_i
```
Why this over `n.to_s.length`?

rsutphin replied:
When the question is "how many digits are in a number?", `log` is where my mind goes. The other would be fine also.

hannahwhy commented on lib/ncs_navigator/core/case_cloner.rb:
```diff
@@ -0,0 +1,299 @@
+        FROM participants root
+          INNER JOIN participant_person_links root_persons ON root.id = root_persons.participant_id
+          INNER JOIN participant_person_links other ON root_persons.person_id = other.person_id
+        WHERE root.p_id = '#{@root_p_id}'
```
So, I assume that `CaseCloner` will not be used in a situation where `@root_p_id` would evaluate to e.g. `'; DROP TABLE ...`

That said, a bind parameter would probably be a good idea.

rsutphin replied:
CaseCloner is not used in a situation where the p_id would come from an untrusted user, no, and it probably won't ever be. 

It's probably worthwhile to use a bind parameter for consistency, though it will mean that I can't use `AR::B.connection.select_all` — bind parameters were not supported on the low-level direct query methods last time I checked.

rsutphin committed:
Avoid theoretical SQL injections. #3826.

This class isn't presently used in a way that the p_id is ever
provided by an untrusted user. And even if it is used in the GUI
later, the UI in front of it would probably not directly receive a
p_id to pass to it. But it doesn't hurt to be safe.

rsutphin committed:
Make sure "nil" is in the behavior description. #3826.

rsutphin committed:
Reset to the original importer mode after exception. #3826.

rsutphin commented:
Updated based on @yipdw's comments.

pfriedman commented:
Looks good to me. I wonder if there are going to be other classes to add to `CaseCloner` but I suspect we can handle those as they come.

rsutphin commented:
> I wonder if there are going to be other classes to add to CaseCloner but I suspect we can handle those as they come.

There's some discussion on [the ticket](https://code.bioinformatics.northwestern.edu/issues/issues/show/3826). I think I got all of the current ones, but there certainly might be more in the future.

rsutphin commented:
Merged into `develop`.
